### PR TITLE
Sprint 12 (prepareForReuse, ImagesListServiceDidChange, POST and DELETE, HTTPCookieStorage, UIView.animate, UIView.animateKeyframes, Core Animation)

### PR DESCRIPTION
**Pull request by Dassam**
### **1. Page-by-page data loading for the photo feed implemented:**
- added a structure for decoding a JSON response from Unsplash - PhotoResult (as well as an auxiliary UrlsResult structure for getting fields from urls);
- the Photo structure has been created, which will be used in the UI part of the application;
- when receiving a response from the network, a Photo object is generated for each PhotoResult;
- ISO8601DateFormatter is used to get createdAt: Date from PhotoResult\.createdAt;
- in the table, the createdAt date is substituted in the cell with the photo instead of the current date;
- created new class ImagesListService;
- added property photos: [Photo] to store a sequence of all photos downloaded from the network;
- static notification was added to ImagesListService let DidChangeNotification = Notification.Name(rawValue: "ImagesListServiceDidChange");
- The fetchPhotosNextPage() function is declared in the ImagesListService class, which loads the next page. Wherein -
- the function internally determines the number of the next page to download (the number should not be reported from the outside, as a function parameter);
- if there is a download, then no new network request is created, and the execution of the function is interrupted;
- when new photos are received, the photos array is updated from the main thread, and new photos are added to the end of the array;
- after updating the value of the photos array, the notification ImagesListService.DidChangeNotification is published.
- the updateTableViewAnimated() function was written, which will be called upon notification ImagesListService.DidChangeNotification and will animatedly update the state of the table.

### **2. Implemented Like functionality:**
- the user has the ability to interact with the indicator — a button with heart pictures (UIButton) has been created;
- indicators are located on each picture - the button with the indicator is placed in the table cell;
- the setIsLiked method has been added to the cell, which changes the state (picture) of the like indicator;
- created @IBAction for the tap on like reaction and @IBOutlet for the button;
- handled a network race situation by blocking the UI interface.

### **3. When image in the general list of images is taped, a screen opened where user can view the full-size version of the photo:**
- UIImageView is used to display the image;
- UIScrollView is used to view the picture in scale;
- while the image is loading, the loading loader is shown;
- Implemented field parsing with a link to a full-sized image;
- a transition to a new controller has been implemented to view the selected picture from the list;
- the new controller transfers information about it to display the selected picture;
- Implemented the "Back" button to return to the list of pictures.

### **4. Added log out ability from the account:**
- by clicking on the exit button, showing a warning and requesting confirmation of the exit; when the user confirms logging out of the account, then:
          -     the authToken value is removed from the OAuth2TokenStorage;
          -     web browser cookies are cleared otherwise, when the browser is opend, login and password fields are not shown;
          -     rootViewController is replaced with SplashViewController.

### CODE REVIEW

File: ImageFeedList/Presentation/Login/ImagesList/ImagesListCell.swift [hunk 1]
```diff
@@ -47,6 +45,13 @@ final class ImagesListCell: UITableViewCell {
         return label
     }()
     
+    private lazy var dateFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.locale = Locale(identifier: "RU_ru")
+        formatter.dateFormat = "dd MMMM yyyy"
+        return formatter
+    }()
```
Comment (andrew-kuzn-yndx) [inline ImageFeedList/Presentation/Login/ImagesList/ImagesListCell.swift]: Такие вещи лучше выводить в extension, например) 
dateFormatter может пригодиттьтся в разных частях приложения, и не придется везде его прописывать

File: ImageFeedList/Service/ImagesListService.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+//  ImageFeedList
+//
+//  Created by Dassam on 30.07.2023.
+//
+
+import UIKit
+
+final class ImagesListService {
+    private lazy var dateFormatter = {
+            return ISO8601DateFormatter()
```
Comment (andrew-kuzn-yndx) [inline ImageFeedList/Service/ImagesListService.swift]: Вот снова dateFormatter - лучше вывести в extension

File: ImageFeedList/Service/ImagesListService.swift [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                          largeImageURL: $0.urls.full,
+                          isLiked: $0.likedByUser
+                    )
+                }
+                DispatchQueue.main.async {
+                    self.photos.append(contentsOf: mapPhotos)
+                    NotificationCenter.default.post(name: Self.didChangeNotification, object: self)
+                }
+            case .failure(let error):
+                assertionFailure(error.localizedDescription)
```
Comment (andrew-kuzn-yndx) [inline ImageFeedList/Service/ImagesListService.swift]: Может лучше вызывать не assertionFailure, а показать алерт пользователю?

File: ImageFeedList/Service/ImagesListService.swift (hunk 2 above)
Comment (Dassam) [reply ImageFeedList/Service/ImagesListService.swift]: Тут обратная операция мы парсим из строки Дату 
По этому не понятно как сделать extension в данном случае((

File: ImageFeedList/Presentation/Login/ImagesList/ImagesListCell.swift (hunk 1 above)
Comment (Dassam) [reply ImageFeedList/Presentation/Login/ImagesList/ImagesListCell.swift]: Done

File: ImageFeedList/Service/ImagesListService.swift (hunk 3 above)
Comment (Dassam) [reply ImageFeedList/Service/ImagesListService.swift]: Done

**Pull request merged**